postgis/lwgeom_hexgrid.c

```diff
@@ -0,0 +1,360 @@
+Datum ST_ShapeGrid(PG_FUNCTION_ARGS)
+{
+	int i, j;
+	FuncCallContext *funcctx;
+	MemoryContext oldcontext, newcontext;
+
+	GSERIALIZED *gbounds;
+	GeometryGridState *state;
+
+	LWGEOM *lwgeom;
+	bool isnull[3] = {0,0,0}; /* needed to say no value is null */
+	Datum tuple_arr[3]; /* used to construct the composite return value */
+	HeapTuple tuple;
+	Datum result; /* the actual composite return value */
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		const char *func_name;
+		double bounds_width, bounds_height;
+		int gbounds_is_empty;
+		GBOX bounds;
+		double size;
+		funcctx = SRF_FIRSTCALL_INIT();
+
+		/* get a local copy of what we're doing a dump points on */
+		gbounds = PG_GETARG_GSERIALIZED_P(1);
+		size = PG_GETARG_FLOAT8(0);
+
+		gbounds_is_empty = (gserialized_get_gbox_p(gbounds, &bounds) == LW_FAILURE);
+		bounds_width = bounds.xmax - bounds.xmin;
+		bounds_height = bounds.ymax - bounds.ymin;
+
+		/* quick opt-out if we get nonsensical inputs  */
+		if (size <= 0.0 || gbounds_is_empty ||
+		    bounds_width <= 0.0 || bounds_height <= 0.0)
+		{
+			funcctx = SRF_PERCALL_SETUP();
+			SRF_RETURN_DONE(funcctx);
+		}
+
+		newcontext = funcctx->multi_call_memory_ctx;
+		oldcontext = MemoryContextSwitchTo(newcontext);
+
+		/* Create function state */
+		state = lwalloc(sizeof(GeometryGridState));
+		state->cell_current = 0;
+		state->bounds = bounds;
+		state->srid = gserialized_get_srid(gbounds);
+		state->size = size;
+
+		func_name = get_func_name(fcinfo->flinfo->fn_oid);
+		if (strcmp(func_name, "st_hexagongrid") == 0)
+		{
+			state->cell_shape = SHAPE_HEXAGON;
+			state->cell_count = hexagon_grid_size(size, &state->bounds,
+			                                      &state->cell_width, &state->cell_height);
+		}
+		else if (strcmp(func_name, "st_squaregrid") == 0)
+		{
+			state->cell_shape = SHAPE_SQUARE;
+			state->cell_count = square_grid_size(size, &state->bounds,
+			                                     &state->cell_width, &state->cell_height);
+		}
+		else
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				errmsg("%s called from unsupported functional context '%s'", __func__, func_name)));
+		}
+
+		funcctx->user_fctx = state;
+
+		/* get tuple description for return type */
+		if (get_call_result_type(fcinfo, 0, &funcctx->tuple_desc) != TYPEFUNC_COMPOSITE)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				errmsg("set-valued function called in context that cannot accept a set")));
+		}
+
+		BlessTupleDesc(funcctx->tuple_desc);
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	/* stuff done on every call of the function */
+	funcctx = SRF_PERCALL_SETUP();
+	newcontext = funcctx->multi_call_memory_ctx;
+
+	/* get state */
+	state = funcctx->user_fctx;
+
+	/* Stop when we've used up all the grid squares */
+	if (state->cell_current >= state->cell_count)
+	{
+		SRF_RETURN_DONE(funcctx);
+	}
+
+	i = state->cell_current % state->cell_width;
+	j = state->cell_current / state->cell_width;
```

Comment:
Should j be based on `cell_height` and not `cell_width` in case they are different?

Reply:
I really wanted to avoid the non-square case, until someone asked for it.